internal/rest/http_client_test.go

```diff
@@ -0,0 +1,195 @@
+	client.httpClient.Transport = fn
+}
+
+func TestHTTPClient_Get(t *testing.T) {
+	type args struct {
+		serverHandler roundTripFunc
+		url           string
+		headers       []HTTPHeader
+	}
+	tests := []struct {
+		err  *Error
+		args args
+		name string
+		want []byte
+	}{
+		{
+			name: "happy path, return success and response",
+			args: args{
+				url: "localhost:8080/some/path",
+				headers: []HTTPHeader{
+					NewHTTPHeader("Some-Header", "value"),
+				},
+				serverHandler: func(req *http.Request) *http.Response {
+					header := req.Header.Get("Some-Header")
+					assert.Equal(t, "value", header)
+					assert.Equal(t, "localhost:8080/some/path", req.URL.String())
+					return &http.Response{
+						StatusCode: 200,
+						// Send response to be tested
+						Body: ioutil.NopCloser(bytes.NewBufferString("OK")),
+						// Must be set to non-nil value or it panics
+						Header: make(http.Header),
+					}
+				},
+			},
+		},
+		{
+			name: "return error from server, status code <= 500",
+			args: args{
+				serverHandler: func(req *http.Request) *http.Response {
+					return &http.Response{
+						StatusCode: 500,
+						// Send response to be tested
+						Body: ioutil.NopCloser(bytes.NewBufferString("FAIL")),
+						// Must be set to non-nil value or it panics
+						Header: make(http.Header),
+					}
+				},
+			},
+			err: NewError(500, "FAIL"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := NewHTTPClient()
+			setTransport(tt.args.serverHandler, c)
+			ctx := context.Background()
+			resp, err := c.Get(ctx, tt.args.url, tt.args.headers...)
+			if tt.err != nil {
+				assert.Equal(t, tt.err, err)
+				return
+			}
+			assert.Nil(t, err)
+			assert.Equal(t, "OK", string(resp))
+		})
+	}
+}
+
+func TestHttpClient_Retry(t *testing.T) {
+	c := NewHTTPClient()
+	ctx := context.Background()
+	retry := 1
+	setTransport(func(req *http.Request) *http.Response {
+		resp := &http.Response{
+			StatusCode: 200,
+			// Send response to be tested
+			Body: ioutil.NopCloser(bytes.NewBufferString("OK")),
+			// Must be set to non-nil value or it panics
+			Header: make(http.Header),
+		}
+		if retry < 3 {
+			// retry range
+			resp.StatusCode = 500
+			retry++
+		}
+		return resp
+	}, c)
+	resp, err := c.Get(ctx, "somehost:8080")
+	assert.Nil(t, err)
+	assert.Equal(t, "OK", string(resp))
+	assert.Equal(t, 3, retry)
+	// test non-retryable status code
+	retry = 0
+	setTransport(func(req *http.Request) *http.Response {
+		resp := &http.Response{
+			StatusCode: 400,
+			Body:       ioutil.NopCloser(bytes.NewBufferString("OK")),
+			Header:     make(http.Header),
+		}
+		retry++
+		return resp
+	}, c)
+	resp, err = c.Get(ctx, "somehost:8080")
+	assert.NotNil(t, err)
+	assert.Nil(t, resp)
+	assert.Equal(t, 1, retry)
+}
+
+// type for json tests
+type employee struct {
+	Name string `json:"name"`
+	Job  string `json:"job"`
+}
+
+func TestHTTPClient_GetJSONObject(t *testing.T) {
```

Comment:
Minor: Maybe we can mock HTTPClient.get() to return an err to test one more branch of "GetJSONObject", which will test  "GetJSONObject returns the error HTTPClient.get() returns"

Reply:
Great idea! Also added similar tests for GetJSONArray at https://github.com/hazelcast/hazelcast-go-client/pull/703/commits/48594298688387b76ed9fd3280296350ca7355fb